src/drivers/uavcan/sensors/battery.cpp

```diff
@@ -69,36 +72,72 @@ UavcanBatteryBridge::battery_sub_cb(const uavcan::ReceivedDataStructure<uavcan::
 
 	battery.timestamp = hrt_absolute_time();
 	battery.voltage_v = msg.voltage;
-	battery.voltage_filtered_v = battery.voltage_v;
+	battery.voltage_filtered_v = msg.voltage;
 	battery.current_a = msg.current;
-	battery.current_filtered_a = battery.current_a;
+	battery.current_filtered_a = msg.current;
 	// battery.average_current_a = msg.;
-	// battery.discharged_mah = msg.;
 
-	// between 0 and 1
-	if (msg.full_charge_capacity_wh > 0) {
-		battery.remaining = msg.remaining_capacity_wh / msg.full_charge_capacity_wh;
-
-	} else {
-		battery.remaining = 0;
-
-	}
+	sumDischarged(battery.timestamp, battery.current_a);
+	battery.discharged_mah = _discharged_mah;
 
+	battery.remaining = msg.state_of_charge_pct / 100.0f; // between 0 and 1
 	// battery.scale = msg.; // Power scaling factor, >= 1, or -1 if unknown
-	battery.temperature = msg.temperature;
+	battery.temperature = msg.temperature - 273.15f; // Kelvin to Celcius
 	// battery.cell_count = msg.;
-	// battery.voltage_cell_v[4] = msg.;
-	// battery.max_cell_voltage_delta = msg.;
+	battery.connected = true;
+	battery.system_source = msg.status_flags & uavcan::equipment::power::BatteryInfo::STATUS_FLAG_IN_USE;
+	// battery.priority = msg.;
 	battery.capacity = msg.full_charge_capacity_wh;
 	// battery.cycle_count = msg.;
 	// battery.run_time_to_empty = msg.;
 	// battery.average_time_to_empty = msg.;
 	battery.serial_number = msg.model_instance_id;
-	battery.connected = true;
-	battery.system_source = msg.status_flags & uavcan::equipment::power::BatteryInfo::STATUS_FLAG_IN_USE;
-	// battery.priority = msg.;
+	battery.id = msg.getSrcNodeID().get();
+
+	// battery.voltage_cell_v[0] = msg.;
+	// battery.max_cell_voltage_delta = msg.;
+
 	// battery.is_powering_off = msg.;
-	// battery.warning = msg.;
+
+	determineWarning(battery.remaining);
+	battery.warning = _warning;
 
 	publish(msg.getSrcNodeID().get(), &battery);
 }
+
+void
+UavcanBatteryBridge::sumDischarged(hrt_abstime timestamp, float current_a)
+{
+	// Not a valid measurement
+	if (current_a < 0.f) {
+		// Because the measurement was invalid we need to stop integration
+		// and re-initialize with the next valid measurement
+		_last_timestamp = 0;
+		return;
+	}
+
+	// Ignore first update because we don't know dt.
+	if (_last_timestamp != 0) {
+		const float dt = (timestamp - _last_timestamp) / 1e6;
+		// mAh since last loop: (current[A] * 1000 = [mA]) * (dt[s] / 3600 = [h])
```

Comment:
Did you consider using the battery lib for this (I haven't looked in detail)?

Reply:
I did not consider it. But now that I look, this could inherit from Battery and use those functions.

Reply:
I don't think it is possible with the current design of the battery lib. Sumdischarged and determineWarning are private members. So I would have to use updateBatteryStatus, but that doesn't have all the fields required. So I could modify the battery lib to work in this situation, or leave these two functions duplicated into this driver. 
Also, UavcanCDevSensorBridgeBase already has its structure of publishing to uorb.